MOB-1747: Implement bookmark export / import

kimar opened the pull request:
# Rationale

This PR introduces a bookmark import/export functionality. Please consider all subtasks of [MOB-1747](https://ecosia.atlassian.net/browse/MOB-1747) when reviewing this PR.

# Demo

| NTP Card | "New Badge" | Tooltip | Entry point for import/export file pickers |
|---|---|---|---|
|  ![Simulator Screenshot - iPhone 14 Pro - 2023-05-10 at 15 03 56](https://github.com/ecosia/ios-browser/assets/126418/86865771-42d6-4342-8be4-32a87a34fb65)| ![Simulator Screenshot - iPhone 14 Pro - 2023-05-10 at 14 37 37](https://github.com/ecosia/ios-browser/assets/126418/2c129153-0a89-4968-a0fe-c3808db1bb1d) | ![Simulator Screenshot - iPhone 14 Pro - 2023-05-10 at 14 37 44](https://github.com/ecosia/ios-browser/assets/126418/9fc17627-ac0b-4bb9-87c3-1e8b59ab488d) |![Simulator Screenshot - iPhone 14 Pro - 2023-05-10 at 14 37 49](https://github.com/ecosia/ios-browser/assets/126418/34f1541f-89a7-45dc-97c1-f8fb97c23cdf)|

# ToDo

* Make sure to remove https://github.com/ecosia/ios-browser/pull/437/files/76f51ecd25ef996b9d6f450a29d5fcb45a70be3c#diff-78a8a19706dbd2a4425dd72bdab0502ed7a2cef16365ab7030a5a0588927bf47R95 before merging, once QA on this PR is done.

kimar commented on Client/Ecosia/UI/NTP/Cells/NTPBookmarkNudgeCell.swift:
```diff
@@ -0,0 +1,184 @@
+        button.translatesAutoresizingMaskIntoConstraints = false
+        button.setImage(UIImage(named: "xmark"), for: .normal)
+//        button.contentVerticalAlignment = .fill
+//        button.contentHorizontalAlignment = .fill
```
Note to self: This needs to be removed @kimar

d4r1091 commented on Client/Ecosia/Extensions/HomepageViewController+Ecosia.swift:
```diff
@@ -91,3 +99,16 @@ extension HomepageViewController: YourImpactDelegate {
+extension HomepageViewController: NTPBookmarkNudgeViewDelegate {
+    func nudgeCellOpenBookmarks() {
+        homePanelDelegate?.homePanelDidRequestToOpenLibrary(panel: .bookmarks)
+        User.shared.hideBookmarksNTPNudgeCard()
```
I'm sure there is some clear need to decorate the `User`'s `shared` instance with the ability to interact with bookmarks or cards. Asking as I need to become more familiar with the user's instance and its abilities overall 🙏 .
This comment is valid for all the following unfamiliar (to me) behaviors belonging to the user's shared instance.

kimar replied:
This was my initial assumption, that `User.shared` keeps this state, but now that you're bringing this up it would be good to get some confirmation on this. Basically I'm keeping different states because those different elements are shown at different points in time due to different criteria being required for them to show up.

d4r1091 replied:
I guess that's the question regarding the `User.shared` @kimar @ecotopian

ecotopian replied:
Historically we have been using the `User.shared` struct to persist this kind of application state whether popups/banners should show or have been shown. So it's totally relatable that Marcus followed that path and I don't see an imminent issue with this. 

In the long run I can imagine that the User struct gets bloated and harder to maintain and oversee. If we agree that this is the case, it can make sense to split up/extract the data which is less user related and think of a different persistency mechanism. But this should not be part of Marcus' tasks here. Instead we can assess this dedicatedly.

d4r1091 commented on Client/Ecosia/L10N/String.swift:
```diff
@@ -182,5 +182,20 @@ extension String {
+        case exportingBookmarks = "Exporting Bookmarks…"
+        case importingBookmarks = "Importing Bookmarks…"
+        case importedBookmarkFolderName = "Imported Bookmarks (%@)"
+        case bookmarkImportHelpscoutURL = "https://ecosia.helpscoutdocs.com/search?query=Bookmark"
```
Does the URL need to stay within the localization `String.swift` file? Trying to get my head around possible string interpolation and utilization within a `UITextView` or `UILabel` maybe. But even so, this URL feels like it may be better into another place? Or even better start to adapt another convenient wrapping `struct` for those things and include the URL in `MultiplyYourImpact:462` 🤷‍♂️ 💭

kimar replied:
This will be removed, after discussing the matter it turns out this will be a single URL which will adjust to the user's language preference of the browser accordingly.

d4r1091 commented on Client/Ecosia/UI/EmptyBookmarksView.swift:
```diff
@@ -0,0 +1,179 @@
+final class EmptyBookmarksView: UIView, NotificationThemeable {
+
+    required init?(coder: NSCoder) {
+        assertionFailure("This view is only supposed to be instantiated programmatically")
```
Nice check 👌

d4r1091 commented on Client/Ecosia/UI/EmptyBookmarksView.swift:
```diff
@@ -0,0 +1,179 @@
+        let label = UILabel()
+        label.text = .localized(.noBookmarksYet)
+        label.textAlignment = .center
+        label.font = UIFontMetrics(forTextStyle: .body).scaledFont(for: .systemFont(ofSize: 17, weight: .semibold))
```
I've noticed that the codebase utilizes some `UX` wrapper at the top of the classes to put every hardocded dimension / UI-related stuff in one place to facilitate updates if needed.
There may be other places where the same suggestion may come up so I won't comment there too 👍 .

kimar replied:
Good catch, thx. Done.

d4r1091 commented on Client/Ecosia/UI/EmptyBookmarksView.swift:
```diff
@@ -0,0 +1,179 @@
+        return button
+    }()
+    
+    private let learnMoreHandler: () -> Void
+    private let importBookmarksHandler: () -> Void
```
I've noticed that for the NTP we have utilized the delegate pattern while, for this view, we implement _callbacks_ ( to elevate the even-driven approach with the " `onSomethingHappened` ", which I like more 😇 )...I'm conscious there is no clear (and shared) pattern within the team in terms of "code style pattern" (yet perhaps 😅  ), but I wonder whether there is a specific reason for choosing this methodology here.

kimar replied:
Yep, makes sense to align this I think. I just picked the function closure callbacks because I find them less verbose but also I think it makes sense to align with the majority of the codebase.

d4r1091 commented on Client/Ecosia/UI/EmptyBookmarksView.swift:
```diff
@@ -0,0 +1,179 @@
+        // title
+        containerStackView.addArrangedSubview(titleLabel)
+        
+        // space between title and first section
```
Thanks for the explanation of those ❤️

d4r1091 commented on Client/Ecosia/UI/NTP/Cells/NTPBookmarkNudgeCell.swift:
```diff
@@ -0,0 +1,182 @@
+
+final class NTPBookmarkNudgeCell: UICollectionViewCell, NotificationThemeable, ReusableCell {
+    
+    private enum UX {
```
Here the approach of `UX` has been followed for some of the elements👌 - for consistency, I'd make some effort to do the same perhaps 🙏

d4r1091 commented on Client/Ecosia/UI/NTP/Cells/NTPBookmarkNudgeCell.swift:
```diff
@@ -0,0 +1,182 @@
+        openBookmarksButton.addTarget(self, action: #selector(handleOpenBookmarks), for: .touchUpInside)
+    }
+    
+    func applyTheme() {}
```
Is there a reason for having this class conforming to `NotificationThemeable` if the `applyTheme` isn't utilized?

kimar replied:
I've missed the thumbing here, also I just learned about the notification which we subscribe to to update the themes, added this. thx!

d4r1091 commented on Client/Ecosia/UI/NTP/Cells/NTPBookmarkNudgeCell.swift:
```diff
@@ -0,0 +1,182 @@
+    }
+}
+
+private class NTPBookmarkNudgeCellBadge: UIView {
```
Apologize in advance if you were planning to do this anyway.
Leaving it as "reminder". What do you think about moving this class elsewhere?

```suggestion
private final class NTPBookmarkNudgeCellBadge: UIView {
```

kimar replied:
Yep, totally makes sense to me! Updating.

d4r1091 commented on Client/Ecosia/UI/NTP/NTPBookmarkNudgeViewModel.swift:
```diff
@@ -0,0 +1,85 @@
+    func nudgeCellDismiss()
+}
+
+class NTPBookmarkNudgeViewModel {
```
```suggestion
final class NTPBookmarkNudgeViewModel {
```

d4r1091 commented on Client/Ecosia/UI/NTP/NTPBookmarkNudgeViewModel.swift:
```diff
@@ -0,0 +1,85 @@
+import Shared
+import Core
+
+protocol NTPBookmarkNudgeViewDelegate: AnyObject {
```
I believe you got the almost "obsessive" approach I've got when it comes to separate files 😇 😅 .
Wondering whether this could be moved into a separate one unless strictly needed here

kimar replied:
Yep! 😄  Done!

d4r1091 commented on Client/Ecosia/UI/NTP/NTPTooltip.swift:
```diff
@@ -98,6 +133,26 @@ final class NTPTooltip: UICollectionReusableView, NotificationThemeable {
+        linkButton.isHidden = true
+    }
+    
+    func setLink(_ text: String) {
```
Make sense to have a string as we are operating with an attributed one within the function. Wonder whether it would sound more consistent to pass a URL rather than a `text: String` 💭 . WDYT?

d4r1091 commented on Client/Frontend/Browser/SimpleToast.swift:
```diff
@@ -19,8 +19,22 @@ struct SimpleToastUX {
 
-struct SimpleToast {
-    func showAlertWithText(_ text: String, image: String, bottomContainer: UIView) {
+class SimpleToast {
```
If possible, unless inherited somewhere 🙏 
```suggestion
final class SimpleToast {
```

d4r1091 replied:
Nice refactor 👌 👏

d4r1091 commented on Client/Frontend/Library/Bookmarks/BookmarksExchange.swift:
```diff
@@ -0,0 +1,186 @@
+    func `import`(from url: URL, in viewController: UIViewController) async throws
+}
+
+class BookmarksExchange: BookmarksExchangable {
```
```suggestion
final class BookmarksExchange: BookmarksExchangable {
```

d4r1091 commented on Client/Frontend/Library/Bookmarks/BookmarksExchange.swift:
```diff
@@ -0,0 +1,186 @@
+            dismissAfter: nil,
+            bottomInset: view.layoutMargins.bottom
+        )
+
+
```
```suggestion

```

d4r1091 commented on Client/Frontend/Library/Bookmarks/BookmarksPanel.swift:
```diff
@@ -562,6 +649,60 @@ extension BookmarksPanel {
         }
     }
+    
+    func showMoreDialog() {
```
Just double checking whether the functions here can be made `private` 🙏

kimar replied:
Yep! Done.

d4r1091 reviewed (commented):
Awesome work @kimar 👏 🎉 🌮 . I performed a preliminary look and left some comments, mostly explanations and double-checking, to better understand the flow and hopefully gather a well-rounded picture of how the components interact.
Huge thanks for the refactor you've done alongside implementing this feature 💪 . Solid work!

kimar commented on Client/Ecosia/Extensions/HomepageViewController+Ecosia.swift:
```diff
@@ -65,6 +65,14 @@ extension HomepageViewController: NTPTooltipDelegate {
     }
+    
+    func ntpTooltipCloseTapped(_ tooltip: NTPTooltip?) {
+        // no-op
```
Yes this makes sense, the no-op just indicates that this method is intentionally left blank. I'll split the delegate so it doesn't need to be declare at all, if not desired.

kimar replied:
Hm I went with the default implementations in a protocol extension, otherwise I would have to declare multiple delegate variables for the distance delegates as I can't constraint one variable to multiple delegates which are optional. (Only all optional or non but not every individual one).

lucaschifino commented on Client/Frontend/Library/Bookmarks/BookmarksPanelViewModel.swift:
```diff
@@ -60,10 +68,73 @@ class BookmarksPanelViewModel {
+            do {
+                let bookmarks = try await getBookmarksForExport()
+                try await bookmarksExchange.export(bookmarks: bookmarks, in: viewController, barButtonItem: viewController.moreButton)
+                await notifyExportDone(nil)
```
Looks like `notifyExportDone` executes the received `onDone` on main, I guess the method was split to use `@MainActor`, but how about using it directly this to reduce code:
```suggestion
            Task { @MainActor in
                onDone(nil)
            }
```
I believe this would reach the same behaviour, but let me know if I'm missing something 🙏
(This also applies for the usage inside the `catch` below)

kimar replied:
Yes it does indeed, I actually thought my version is less verbose because it has less duplication, but effectively you might be right. Let me check...

kimar replied:
Hm, I don't know, tbh. I would prefer:

```swift

                await notifyExportDone(nil)
            } catch {
                await notifyExportDone(error)
            }
```

over

```swift
                Task { @MainActor in onExportDoneHandler?(nil) }
            } catch {
                Task { @MainActor in onExportDoneHandler?(error) }
            }
```

even if there is one method more, just for the sake of having one less closure, but maybe @d4r1091 has another input / opinion about this?

lucaschifino replied:
Your point is also valid, feel free to leave as you prefer as well 😁

I think I would still find the below cleaner since it gets rid of the method and the intermediate closure variable, but this is also personal opinion since we don't have a defined style, so feel free to leave it like it is 🙂
```
                Task { @MainActor in onDone(nil) }
            } catch {
                Task { @MainActor in onDone(error) }
            }
```

Also noticed the `handlePickedUrl` method further down this file has a similar style, so if you'd like you can try and match both to keep the same in this file.

kimar replied:
Ha! You got me with this one 😄  yes I agree that #1 priority should be to be consistent here, I'm fine with both solutions, I'm leaning towards separate method it `@MainActor` annotation just for the sake of removing one closure, but I'M also fine with your proposed solution.

lucaschifino replied:
No worries, you can keep the one you are leaning towards then, both are good alternatives indeed 😁

lucaschifino reviewed (commented):
Nicely done 👏 Indeed the code is very readable for a complex feature, even considering having to integrate into existing bookmark code from firefox 🙏

Added some small questions/comments to take a look at

kimar commented on Client/Ecosia/UI/EmptyBookmarksView.swift:
```diff
@@ -0,0 +1,179 @@
+
+final class EmptyBookmarksView: UIView, NotificationThemeable {
+
+    required init?(coder: NSCoder) {
```
Very good point, I'll update this.

kimar commented on Client/Ecosia/UI/EmptyBookmarksView.swift:
```diff
@@ -0,0 +1,179 @@
+        addSection(imageNamed: "exportShare", text: .localized(.bookmarksEmptyViewItem1))
+        
+        let buttonStackViewSpacer = UIView.build {
+            $0.translatesAutoresizingMaskIntoConstraints = false
```
Yes, you're absolutely right! Adjusted this here and there.

kimar commented on Client/Ecosia/UI/EmptyBookmarksView.swift:
```diff
@@ -0,0 +1,179 @@
+        learnMoreButton.addTarget(self, action: #selector(onLearnMoreTapped), for: .touchUpInside)
+        importBookmarksButton.addTarget(self, action: #selector(onImportTapped), for: .touchUpInside)
+
+        applyTheme()
```
Yes, you're absolutely right, updated this. Thx! 🙏

kimar commented on Client/Frontend/Browser/BrowserViewController.swift:
Yes I think this makes sense. I'll add comments regarding my changes.

kimar commented on Client/Frontend/Library/Bookmarks/BookmarksExchange.swift:
Right, I must have missed this, will move it. thx! 🚀

kimar commented on Client/Frontend/Library/Bookmarks/BookmarksExchange.swift:
```diff
@@ -0,0 +1,186 @@
+        self.profile = profile
+    }
+    
+    @MainActor
```
Thank you 😺

kimar commented:
@d4r1091 @lucaschifino thank you for your reviews! I've addressed (most of) the feedback. The thing regarding the `User.shared` might need some further clarification wether this is the ideal to store this state, at the time of writing this code to me it seemed most appropriate. Maybe @ecotopian has some input later as well.

lucaschifino reviewed (approved):
LGTM 👏 Thank you for addressing the comments 🙏

ecotopian commented on Client/Ecosia/Bookmarks/BookmarksExchange.swift:
```diff
@@ -0,0 +1,187 @@
+            switch bookmark {
+            case let .folder(title, children, _):
+                let subParentGuid = try await createFolder(parentGUID: parentGUID, title: title)
+                try await processBookmarks(children, parentGUID: subParentGuid)
```
neat recursion 👨‍🔬

kimar replied:
Thank you :-)

ecotopian commented on Client/Ecosia/UI/EmptyBookmarksView.swift:
```diff
@@ -0,0 +1,208 @@
+    private let titleLabel: UILabel = {
+        let label = UILabel()
+        label.text = .localized(.noBookmarksYet)
+        label.textAlignment = .center
```
Following up on other comment I'd expect to find `label.adjustsFontForContentSizeCategory = true` here

ecotopian replied:
also on the following labels/buttons

ecotopian commented on Client/Frontend/Browser/SimpleToast.swift:
```diff
@@ -19,8 +19,23 @@ struct SimpleToastUX {
 
-struct SimpleToast {
-    func showAlertWithText(_ text: String, image: String, bottomContainer: UIView) {
+final class SimpleToast {
```
Not so simple anymore, hum 😉 ? But I guess that's how it goes when we add features to it 😆

ecotopian commented on Client/Frontend/Browser/SimpleToast.swift:
```diff
@@ -69,6 +94,12 @@ struct SimpleToast {
+    
+    func dismiss() {
+        DispatchQueue.main.async {
+            self.toastView.map(self.dismiss)
```
It took me some time understanding this logic as I haven't seen that there's a second method called `dismiss` but with the view parameter 🤔 What do you think of making the difference more opaque by renaming one of the methods?

ecotopian commented on Client/Frontend/Library/Bookmarks/BookmarksPanel.swift:
```diff
@@ -20,6 +21,7 @@ class BookmarksPanel: SiteTableViewController, LibraryPanel, CanRemoveQuickActio
     struct UX {
         static let FolderIconSize = CGSize(width: 24, height: 24)
         static let RowFlashDelay: TimeInterval = 0.4
+        static let BookmarksHelpScoutUrlString = "https://ecosia.helpscoutdocs.com/article/458-import-export-bookmarks"
```
Usually we have those kind of URLs in [Core/Environment.swift](https://github.com/ecosia/ios-core/blob/main/Sources/Environment/Environment.swift) to have a central overview of the used URLs. Would you mind moving it there? 🙏

kimar replied:
Oh, that sounds reasonable, I'll move it there then.

ecotopian reviewed (commented):
Really awesome work @kimar!! 🚀🚀🚀 

I am really overwhelmed how many different places had to be altered to finally implement this feature. I guess you learnt a lot how the different components interact 💡 

I only have two minor doubts. One regarding autoscaling fonts in the views that you added, and one regarding the URL constant. But nothing major 👏

kimar commented on Client/Ecosia/Bookmarks/BookmarksExchange.swift:
```diff
@@ -0,0 +1,187 @@
+        )
+
+        do {
+            let html = try String(contentsOf: url)
```
Yes this totally makes sense :-) I'm going to use `URLSession` for this as I haven't seen a native way of `String` for loading the data async.

kimar commented on Client/Ecosia/UI/EmptyBookmarksView.swift:
```diff
@@ -0,0 +1,208 @@
+final class EmptyBookmarksView: UIView, NotificationThemeable {
+    
+    private enum UX {
+        static let TitleLabelFont = UIFontMetrics(forTextStyle: .body).scaledFont(for: .systemFont(ofSize: 17, weight: .semibold))
```
Yes it does, but that said, I think those two are doing slightly different things. The scaledFont mainly deals with dynamic type where adjustsFontForContentSizeCategory comes into effect when the sizer category changes, e.g. when the app is used on an ipad and is being resized. I'm going to add this as well. 👍

kimar commented on Client/Ecosia/UI/NTP/NTPTooltip.swift:
Thank you...

ecotopian reviewed (approved):
Green light from my side 🟢 🚀 🙌

kimar commented:
I've addressed the comments in this PR, however right now main is incompatible with `ios-core`main, once https://github.com/ecosia/ios-browser/pull/451/files#diff-272083409fb6beac51ed4b7e1a5bd6931036ca1fd980e88f34a0f74c57eaf23dR30 is merged I'll update this PR and then we can merge it.

kimar commented:
Thank you @d4r1091 @lucaschifino @ecotopian for the reviews, I've just pushed a fix for the updated dependency.

lucaschifino commented on .circleci/config.yml:
```diff
@@ -92,3 +92,4 @@ workflows:
             branches:
               only: 
                 - /^.*main.*/                                
+                - MOB-1759_Implement_Bookmark_Export_Import
```
I guess this should be removed before merging, but probably still waiting on some feedback from Kunle before that

kimar replied:
Yes right 👍  I would wait until the QA is done with this. Thanks again for pointing this out, let's keep this comment here for visibility. I'll also add this to the PR description.

lucaschifino reviewed (approved):
Looking good 👏🟢

Just a comment with regards to the circle ci temporary config change

d4r1091 reviewed (approved):
🔥 !

kimar commented on Client/Application/AppDelegate.swift:
```diff
@@ -31,7 +31,8 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
 
         self.window = UIWindow(frame: UIScreen.main.bounds)
-
+        window?.tintColor = .theme.ecosia.primaryBrand
```
@ecotopian @d4r1091 @lucaschifino I've set the global tint color so e.g. the file picking dialog also picks it up, but not sure if there might be any unintended side-effect on other areas. Just adding this comment to raise awareness.

ecotopian replied:
I can't give a guarantee to exclude side effects, but I think that the strategy is generally okay to tint system like dialogues. Still I think you need to add the functionality that it reassigns the color on theme changes.

kimar replied:
Done ✅

ecotopian commented on Client/Ecosia/MMP/MMP.swift:
```diff
@@ -23,7 +23,6 @@ struct MMP {
                                                   deviceManufacturer: DeviceInfo.manufacturer,
                                                   deviceModel: DeviceInfo.deviceModelName,
                                                   locale: DeviceInfo.currentLocale,
-                                                  country: DeviceInfo.currentCountry,
```
Hmm.. I am wondering why this is in the changes 🤔 @lucaschifino @d4r1091 Can you tell whether this should be removed in `main`?

ecotopian replied:
@kimar can you tell why that change is there?

ecotopian replied:
> Hmm.. I am wondering why this is in the changes 🤔 @lucaschifino @d4r1091 Can you tell whether this should be removed in `main`?

any thoughts @d4r1091 @lucaschifino ?

ecotopian commented on Client/Ecosia/UI/EmptyBookmarksView.swift:
```diff
@@ -0,0 +1,269 @@
+        }
+        
+        let sectionIcon = UIImageView()
+        sectionIcon.tintColor = .theme.ecosia.secondaryText
```
I would guess this line is even obsolete as `applyTheme()` will also set it, right?

ecotopian commented on Client/Ecosia/UI/EmptyBookmarksView.swift:
```diff
@@ -0,0 +1,269 @@
+        let sectionLabel = UILabel()
+        sectionLabel.font = UX.SectionLabelFont
+        sectionLabel.numberOfLines = 0
+        sectionLabel.textColor = .theme.ecosia.secondaryText
```
..also this is overridden in `applyTheme()`, isn't it? Thus I'd generally advocate to remove setting colors outside of `applyTheme()` to have a single point of truth 🧹

kimar replied:
Looks for this to work we actually another fix, as currently the sub-stackview are not looped over, I've fixed this! Thank you.

ecotopian commented on Client/Ecosia/UI/NTP/NTPTooltip.swift:
```diff
@@ -113,15 +156,36 @@ final class NTPTooltip: UICollectionReusableView, NotificationThemeable {
+            tailCenterConstraint.isActive = false
+            tailLeadingConstraint.isActive = true
+        case .trailing:
+            assertionFailure()
```
..wondering if it makes sense to have `.trailing` after all. Apparently using it will have no meaningful effect until it gets implemented 🤔 Did you have a dedicated use case in mind for that?

kimar replied:
Hm interesting, I was pretty sure I only added cases which we actualyl need. I'll remove this!

ecotopian reviewed (approved):
Great work @kimar I am really looking forward to see this live soon 🚀 I just have a few last minor optimisation thoughts. No blockers.

ecotopian reviewed (dismissed):
Looking great now! 🚀
